Sources/BigInt/BigInt.swift

```diff
@@ -0,0 +1,752 @@
+    return result
+  }
+
+  public init?<T>(exactly source: T) where T: BinaryFloatingPoint {
+    if (source.isNaN || source.isInfinite) ||
+      (source.rounded(.towardZero) != source) {
+      return nil
+    }
+
+    self.init(source)
+  }
+
+  @usableFromInline
+  internal init(words: Words) {
+    self.words = words
+  }
+
+  public static func += (lhs: inout BigInt, rhs: BigInt) {
+    if lhs.words.count == 1, rhs.words.count == 1 {
+      let lhsWord = lhs.words[0]
+      let rhsWord = rhs.words[0]
+
+      let (result, isOverflow) = lhsWord.addingReportingOverflow(rhsWord)
+
+      if !isOverflow {
+        lhs.words[0] = result
+        return
+      }
+      let knownNegativeResult = lhsWord > Int.max && rhsWord > Int.max
+
+      if lhsWord > Int.max || rhsWord > Int.max, !knownNegativeResult {
+        // positive + negative is always smaller, so overflow is a red herring
+        lhs.words[0] = result
+        return
+      }
+    }
+
+    var isOverflow = false
+
+    var rhsWords = rhs.words
+
+    lhs.words.append(lhs._isNegative ? UInt.max : 0)
+    rhsWords.append(rhs._isNegative ? UInt.max : 0)
+
+    BigInt._signExtend(lhsWords: &lhs.words, rhsWords: &rhsWords)
+    var temp: UInt = 0
+    for index in 0 ..< lhs.words.count {
+      var carryOverflow = false
+
+      if isOverflow {
+        (temp, carryOverflow) = rhsWords[index].addingReportingOverflow(1)
+      } else {
+        temp = rhsWords[index]
+      }
+
+      (lhs.words[index], isOverflow) = lhs.words[index].addingReportingOverflow(temp)
+
+      isOverflow = carryOverflow || isOverflow
+    }
+
+    BigInt._dropExcessWords(words: &lhs.words)
+  }
+
+  public init(integerLiteral value: Int) {
+    if value >= 0, value < BigInt._digits.count {
+      self = BigInt._digits[value]
+    } else {
+      words = [UInt(bitPattern: value)]
+    }
+  }
+
+  public static var isSigned: Bool { true }
+
+  public var bitWidth: Int { words.count * MemoryLayout<UInt>.size * 8 }
+
+  public var trailingZeroBitCount: Int { words.first?.trailingZeroBitCount ?? 0 }
+
+  @inlinable
+  public static func < (lhs: BigInt, rhs: BigInt) -> Bool {
+    let lhsNegative = lhs._isNegative
+    let rhsNegative = rhs._isNegative
+
+    if lhsNegative && !rhsNegative { return true }
+    if rhsNegative && !lhsNegative { return false }
+
+    if (lhsNegative && rhsNegative) || (!lhsNegative && !rhsNegative) {
+      if lhs.words.count > rhs.words.count {
+        return lhsNegative ? true : false
+      }
+      if lhs.words.count < rhs.words.count {
+        return lhsNegative ? false : true
+      }
+
+      for i in stride(from: lhs.words.count - 1, through: 0, by: -1) {
+        if lhs.words[i] > rhs.words[i] { return false }
+      }
+    }
+
+    return true
+  }
+
+  @inlinable
+  public static func == (lhs: BigInt, rhs: BigInt) -> Bool {
+    lhs.words == rhs.words
+  }
+
+  @inlinable
+  public static func != (lhs: BigInt, rhs: BigInt) -> Bool {
+    !(lhs == rhs)
+  }
```

Comment:
If you implement `Equatable.==(_:_:)` you should also implement `Hashable.hash(into:)` using the same component(s). Or the compiler can synthesize default implementations for both.

I think the `Equatable.!=(_:_:)` can be removed, because it's the same as the [default implementation](https://github.com/apple/swift/blob/746b58e8e1c1ef1ac09d8031875fd2a08b65597c/stdlib/public/core/Equatable.swift#L179-L197).

Reply:
Technically I believe the auto-generated implementation was already doing exactly this, since `words` is the only stored property on `BigInt`. That said, it rarely hurts to be explicit, so I went ahead and added the `hash(into:)` implementation.